phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -931,220 +947,300 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
                         try {
-                            if (cache != null) cache.close();
-                        } finally {
-                            try {
-                                hTable.close();
-                            } catch (IOException e) {
-                                if (sqlE != null) {
-                                    sqlE.setNextException(ServerUtil.parseServerException(e));
-                                } else {
-                                    sqlE = ServerUtil.parseServerException(e);
-                                }
+                            hTable.close();
+                        } catch (IOException e) {
+                            if (sqlE != null) {
+                                sqlE.setNextException(ServerUtil.parseServerException(e));
+                            } else {
+                                sqlE = ServerUtil.parseServerException(e);
                             }
-                            if (sqlE != null) { throw sqlE; }
                         }
+                        if (sqlE != null) { throw sqlE; }
                     }
-                } while (shouldRetry && retryCount++ < 1);
+                }
+            } while (shouldRetry && retryCount++ < 1);
+        }
+    }
+
+    private void filterIndexCheckerMutations(Map<TableInfo, List<Mutation>> mutationMap,
+            Map<TableInfo, List<Mutation>> unverifiedIndexMutations,
+            Map<TableInfo, List<Mutation>> verifiedOrDeletedIndexMutations) throws SQLException {
+        Iterator<Entry<TableInfo, List<Mutation>>> mapIter = mutationMap.entrySet().iterator();
+
+        while (mapIter.hasNext()) {
+            Entry<TableInfo, List<Mutation>> pair = mapIter.next();
+            TableInfo tableInfo = pair.getKey();
+            if (IndexUtil.isGlobalIndexCheckerEnabled(connection, tableInfo.getHTableName())) {
+                PTable table = tableInfo.getOrigTableRef().getTable();
+                byte[] emptyCF = SchemaUtil.getEmptyColumnFamily(table);
+                byte[] emptyCQ = EncodedColumnsUtil.getEmptyKeyValueInfo(table).getFirst();
+
+                List<Mutation> mutations = pair.getValue();
+
+                for (Mutation m : mutations) {
+                    if (m == null) {
+                        continue;
+                    }
+                    if (m instanceof Delete) {
+                        Put put = new Put(m.getRow());
+                        put.addColumn(emptyCF, emptyCQ, IndexRegionObserver.getMaxTimestamp(m),
+                                IndexRegionObserver.UNVERIFIED_BYTES);
+                        // The Delete gets marked as unverified in Phase 1 and gets deleted on Phase 3.
+                        addToMap(unverifiedIndexMutations, tableInfo, put);
+                        addToMap(verifiedOrDeletedIndexMutations, tableInfo, m);
+                    } else {
```

Comment:
@gokceni - This will fail with a ClassCastException if it encounters an Append, a third kind of Mutation that acts somewhat like a Put, but reads the existing value and increments / concatenates to it. 

Since Appends also have an addColumn method, you probably just need another else if for them and similar logic to the Put.

Reply:
Oh, and Increment.

Reply:
According to @kadirozde , "Increment and append should not be supported by Immutable tables by definition
We need to support them for mutable tables". I guess he will have a jira for that.